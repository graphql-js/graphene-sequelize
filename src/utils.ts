import * as Sequelize from "sequelize";
import { Registry } from "./registry";

export type Model = Sequelize.Model<any, any>;

export const assertSequelizeModel = (model: any) => {
  if (model instanceof <any>Sequelize.Model) {
    throw new Error(
      `Expected to receive a Sequelize.Model. Received a sequelize instance instead: ${model}.`
    );
  }
  if (!model || model.constructor !== Sequelize.Model.constructor) {
    throw new Error(
      `Expected to receive a Sequelize.Model. Received ${model} instead.`
    );
  }
};

export const assertRegistry = (registry: any) => {
  if (!(registry instanceof Registry)) {
    throw new Error(
      `Expected to receive a Registry instance. Received ${registry} instead.`
    );
  }
};
