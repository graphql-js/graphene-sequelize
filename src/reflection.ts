import { Model, assertSequelizeModel } from "./utils";
import "reflect-metadata";
import * as Sequelize from "sequelize";

export const GRAPHENE_SEQUELIZE_MODEL = "graphene:sequelize:model";

// A utility funciton to get the sequelize model given a type
export const getSequelizeModel = (target: any): Sequelize.Model<any, any> => {
  const metadataType: Model = Reflect.getMetadata(
    GRAPHENE_SEQUELIZE_MODEL,
    target
  );
  if (metadataType) {
    return metadataType;
  }
  throw new Error(
    `The target ${target} have no Sequelize type associated to it.`
  );
};

// An utility function to associate a Sequelize Model
// with the specified target.
// This method takes full advantage of the Reflection API.
export const setSequelizeModel = (
  target: any,
  model: Sequelize.Model<any, any>
): void => {
  // Has the provided type a Sequelize Model?
  // Will fail if not.
  assertSequelizeModel(model);

  // First we check this type have no other model associated with it.
  if (Reflect.hasMetadata(GRAPHENE_SEQUELIZE_MODEL, target)) {
    throw new Error(
      `Type ${String(target)} already have a Sequelize model attached.`
    );
  }
  // We define the type metadata through reflection
  Reflect.defineMetadata(GRAPHENE_SEQUELIZE_MODEL, model, target);
};
