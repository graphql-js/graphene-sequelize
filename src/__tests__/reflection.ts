jest.mock("../utils");
import { Model, assertSequelizeModel } from "../utils";
import { getSequelizeModel, setSequelizeModel } from "./../reflection";
import * as Sequelize from "sequelize";

describe("Registry", () => {
  test("getSequelizeModel / setSequelizeModel works", () => {
    class User {}
    const mockSequelize = new Object();
    setSequelizeModel(User, <Model>mockSequelize);
    expect(assertSequelizeModel).toBeCalledWith(mockSequelize);
    expect(getSequelizeModel(User)).toBe(mockSequelize);
    // expect(registry.getTypeForModel(<any>mockSequelize)).toBe(User);
  });

  test("getSequelizeModel fails if doesn't have any model attached", () => {
    class User {}
    expect(() => getSequelizeModel(User)).toThrowErrorMatchingSnapshot();
    // expect(registry.getTypeForModel(<any>mockSequelize)).toBe(User);
  });

  test("setSequelizeModel fails if model is already set", () => {
    class User {}
    const mockSequelize = new Object();
    setSequelizeModel(User, <Model>mockSequelize);
    expect(() =>
      setSequelizeModel(User, <Model>mockSequelize)
    ).toThrowErrorMatchingSnapshot();
  });
});
