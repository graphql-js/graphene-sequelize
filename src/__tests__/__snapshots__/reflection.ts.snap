// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Registry getSequelizeModel fails if doesn't have any model attached 1`] = `
"The target class User {
        } have no Sequelize type associated to it."
`;

exports[`Registry setSequelizeModel fails if model is already set 1`] = `
"Type class User {
        } already have a Sequelize model attached."
`;
